cpp/src/fil/infer.cu

```diff
@@ -549,24 +679,21 @@ __global__ void infer_k(storage_type forest, predict_params params) {
     }
 
     tree_aggregator_t<NITEMS, leaf_algo> acc(
-      params, (char*)sdata + params.cols_shmem_size(), sdata);
+      params, (char*)sdata + params.cols_shmem_size(), sdata,
+      forest.vector_leaf_);
 
     __syncthreads();  // for both row cache init and acc init
 
     // one block works on NITEMS rows and the whole forest
     for (int j = threadIdx.x; j - threadIdx.x < forest.num_trees();
          j += blockDim.x) {
-      /* j - threadIdx.x < forest.num_trees() is a necessary but block-uniform
-         condition for "j < forest.num_trees()". It lets use __syncthreads()
-         and is made exact below.
-      */
-      if (j < forest.num_trees()) {
-        acc.accumulate(infer_one_tree<NITEMS, leaf_output_t<leaf_algo>::T>(
-                         forest[j], cols_in_shmem ? sdata : block_input,
-                         num_cols, num_input_rows),
-                       j, num_input_rows);
-      }
-      if (leaf_algo == GROVE_PER_CLASS_MANY_CLASSES) __syncthreads();
+      auto pred = j < forest.num_trees()
+                    ? infer_one_tree<NITEMS, leaf_output_t<leaf_algo>::T>(
+                        forest[j], cols_in_shmem ? sdata : block_input,
+                        num_cols, num_input_rows)
+                    : vec<NITEMS, typename leaf_output_t<leaf_algo>::T>();
+
+      acc.accumulate(pred, j, j < forest.num_trees() ? num_input_rows : 0);
```

Comment:
nice encapsulation! this line was ugly
` if (leaf_algo == GROVE_PER_CLASS_MANY_CLASSES) __syncthreads();`